idx/memory/find_cache.go

```diff
@@ -0,0 +1,189 @@
+package memory
+
+import (
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/grafana/metrictank/stats"
+	lru "github.com/hashicorp/golang-lru"
+	"github.com/raintank/schema"
+	log "github.com/sirupsen/logrus"
+)
+
+var (
+	// metric idx.memory.find-cache.hit is a counter of findCache hits
+	findCacheHit = stats.NewCounterRate32("idx.memory.find-cache.hit")
+	// metric idx.memory.find-cache.miss is a counter of findCache misses
+	findCacheMiss = stats.NewCounterRate32("idx.memory.find-cache.miss")
+
+	findCacheSize            = 1000
+	findCacheInvalidateQueue = 100
+	findCacheBackoff         = time.Minute
+)
+
+type FindCache struct {
+	cache map[uint32]*lru.Cache
+	size  int
+	sync.RWMutex
+	newSeries map[uint32]chan struct{}
+	backoff   map[uint32]time.Time
+}
+
+func NewFindCache(size int) *FindCache {
+	fc := &FindCache{
+		cache:     make(map[uint32]*lru.Cache),
+		size:      size,
+		newSeries: make(map[uint32]chan struct{}),
+		backoff:   make(map[uint32]time.Time),
+	}
+	return fc
+}
+
+func (c *FindCache) Get(orgId uint32, pattern string) ([]*Node, bool) {
+	c.RLock()
+	cache, ok := c.cache[orgId]
+	c.RUnlock()
+	if !ok {
+		findCacheMiss.Inc()
+		return nil, false
+	}
+	nodes, ok := cache.Get(pattern)
+	if !ok {
+		return nil, ok
+	}
+	findCacheHit.Inc()
+	return nodes.([]*Node), ok
+}
+
+func (c *FindCache) Add(orgId uint32, pattern string, nodes []*Node) {
+	c.RLock()
+	cache, ok := c.cache[orgId]
+	t := c.backoff[orgId]
+	c.RUnlock()
+	var err error
+	if !ok {
+		// dont init the cache if we are in backoff mode.
+		if time.Until(t) > 0 {
+			return
+		}
+		cache, err = lru.New(c.size)
+		if err != nil {
+			log.Errorf("memory-idx: findCache failed to create lru. err=%s", err)
+			return
+		}
+		c.Lock()
+		c.cache[orgId] = cache
+		c.newSeries[orgId] = make(chan struct{}, findCacheInvalidateQueue)
+		c.Unlock()
+	}
+	cache.Add(pattern, nodes)
+}
+
+func (c *FindCache) Purge(orgId uint32) {
+	c.RLock()
+	cache, ok := c.cache[orgId]
+	c.RUnlock()
+	if !ok {
+		return
+	}
+	cache.Purge()
+}
+
+func (c *FindCache) PurgeAll() {
+	c.RLock()
+	orgs := make([]uint32, len(c.cache))
+	i := 0
+	for k := range c.cache {
+		orgs[i] = k
+		i++
+	}
+	c.RUnlock()
+	for _, org := range orgs {
+		c.Purge(org)
+	}
+}
+
+func (c *FindCache) InvalidateFor(orgId uint32, path string) {
+	c.RLock()
+	cache, ok := c.cache[orgId]
+	c.RUnlock()
+	if !ok || cache.Len() < 1 {
+		return
+	}
+
+	select {
+	case c.newSeries[orgId] <- struct{}{}:
+	default:
+		c.Lock()
+		c.backoff[orgId] = time.Now().Add(findCacheBackoff)
```

Comment:
is the primary reason for this backoff mechanism because we don't want to slow down the startup procedure when MT is adding a lot of new series? or is it expected that this backoff might occasionally also happen during run time? if it happens during run time we might want to get alerted on that, because it will probably cause the find performance to drop, so maybe we should have a metric that records when we decide to backoff...

Reply:
This only affects the index/find right? So, during startup shouldn't it just respond with a service unavailable until it is in a ready state?

I agree if it happens during run time it would probably be useful to get alerted.

Reply:
If we only want to ensure that the startup doesn't get slowed down, then wouldn't it make more sense to just have an `Activate()` method on the cache which MT can call once the startup is complete? And then we don't need the whole backoff mechanism.
I'm a little worried about a scenario where all of a sudden the find latency jumps and it'll take some grepping through the logs to figure out why.

Reply:
@robert-milan  `InvalidateFor()` gets called when metrics get added to the index, so on startup it would be called a lot of times, for no purpose because the cache will be empty: https://github.com/grafana/metrictank/pull/1233/files#diff-051eec914f9c2892a0d9f0d2a4d87336R465

Reply:
The purpose is to accommodate large bursts of new series.  If the findCache is full (default 1000 items), then every new series seen is compared against all 1000 cache patterns.  That is an expensive overhead for ingestion.  So, the moment a large burst of new series is detected (via not being able to process them fast enough), we just disable the cache for a while.

The backoff isn't really needed for startup, as the findCaches aren't initialized until queries are received.  MT instances dont accept queries until they have completed processing their backlog.

The backoff is more designed for customers sending lots of new series, eg. due to customers sending traffic for new datacenters or servers.

Reply:
The cacheHit ration will go to 0 when the cache is disabled, so we dont need any additional metric to track it.   An Info message is logged everytime that the cache is disabled.  So if the cacheHit ratio drops to 0, it would be easy to look at the logs and see why.

Reply:
The reason why `InvalidateFor()` is called on `.add()` is because the newly added metric might need to show up in the results of some common queries, right? So once the backoff time has passed and the cache gets reactivated again, isn't it going to contain stale entries that should have been invalidated during the time it was disabled.